core/src/main/java/com/graphhopper/reader/dem/EdgeSampling.java

```diff
@@ -0,0 +1,72 @@
+/*
+ *  Licensed to GraphHopper GmbH under one or more contributor
+ *  license agreements. See the NOTICE file distributed with this work for
+ *  additional information regarding copyright ownership.
+ *
+ *  GraphHopper GmbH licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except in
+ *  compliance with the License. You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package com.graphhopper.reader.dem;
+
+import com.graphhopper.util.DistanceCalc;
+import com.graphhopper.util.DistanceCalcEarth;
+import com.graphhopper.util.PointList;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Ensures that elevation is sampled along a point list with no more than maxDistance between samples. Works by adding
+ * points along long edges and fetching elevation at each inserted point.
+ *
+ * This uses a simple linear approximation to interpolate between points which should be fine for short distances, but
+ * if support for longer distances is needed or to handle segments crossing the international date line, should use a
+ * more robust algorithm that interpolates along the great circle path between points.
+ */
+public class EdgeSampling {
+    private static final Logger logger = LoggerFactory.getLogger(EdgeSampling.class);
+    private static final double WARN_ON_SEGMENT_LENGTH = DistanceCalcEarth.METERS_PER_DEGREE / 4;
+
+    private EdgeSampling() {}
+
+    public static PointList sample(PointList input, double maxDistance, DistanceCalc distCalc, ElevationProvider elevation) {
+        PointList output = new PointList(input.getSize() * 2, input.is3D());
+        if (input.isEmpty()) return output;
+        int nodes = input.getSize();
+        double lastLat = input.getLat(0), lastLon = input.getLon(0), thisLat, thisLon, thisEle;
+        for (int i = 0; i < nodes; i++) {
+            thisLat = input.getLat(i);
+            thisLon = input.getLon(i);
+            thisEle = input.getEle(i);
+            if (i > 0 && !distCalc.isCrossBoundary(lastLon, thisLon)) {
+                double segmentLength = distCalc.calcDist(lastLat, lastLon, thisLat, thisLon);
+                if (segmentLength > WARN_ON_SEGMENT_LENGTH) {
+                    logger.warn("Edge from " + lastLat + "," + lastLon + " to " + thisLat + "," + thisLon + " is " +
+                            (int)segmentLength + "m long, might cause issues with edge sampling.");
```

Comment:
I am wondering if we should make this warning a bit more explicit / self-explaining or maybe add a note what the user can do about it?

Crossing the date line should not happen, since we check isCrossBoundary a few lines above.

What are other cases where this could happen? For ferries, this might be an issue. For regular roads, not having a node every few kms could indicate a mapping error. For OSM this could mean to double check the roads at the given coordinates and maybe align the roads. 

I haven't tested the PR yet, are there many of these logs? I could imagine that for example in Canada or Australia, this could be an issue.

Reply:
I haven't run this on the entire world so I don't know if it actually comes up. I did not see it running in the US or Germany.  A few options I could see are:
- handle this more gracefully by falling back to great circle interpolation in this case (or when crossing the international date line)
- not interpolating at all if the edge is too long
- or telling the user to add intermediate points to fix data in OpenStreetMap

If we can find an example, it would be good to see what it looks like on a few map providers - routes returned by GraphHopper should align with maps they might be displayed on.

Reply:
Yes I agree, having an example or two would be great. Ok so I think we should keep this in mind and see if any issues pop up when importing a planet file. I will run a test on Australia to see if there are any such warnings popping up.

Reply:
Ok, I cound find some instances of this warning in Australia :).

```
2020-05-06 13:58:47.635 [main] WARN  c.g.reader.dem.EdgeSampling - Edge from -32.68276041782112,120.69303365110682 to -32.87841862913412,120.4256310036547 is 33140m long, might cause issues with edge sampling.
2020-05-06 14:00:00.221 [main] WARN  c.g.reader.dem.EdgeSampling - Edge from -32.99502934504366,120.25418606223582 to -32.99459069210121,119.94914326703123 is 28448m long, might cause issues with edge sampling.
2020-05-06 14:01:07.068 [main] WARN  c.g.reader.dem.EdgeSampling - Edge from -33.58228318949677,139.9640097833576 to -33.297097123349396,140.12294594473272 is 34972m long, might cause issues with edge sampling.
2020-05-06 14:01:21.906 [main] WARN  c.g.reader.dem.EdgeSampling - Edge from -28.882821363571765,130.60277563935762 to -28.853342581987587,130.90639108955244 is 29746m long, might cause issues with edge sampling.
2020-05-06 14:01:27.928 [main] WARN  c.g.reader.dem.EdgeSampling - Edge from -27.22052787737238,153.12535546255162 to -27.079826267357756,153.36294665998847 is 28238m long, might cause issues with edge sampling.
2020-05-06 14:01:55.286 [main] WARN  c.g.reader.dem.EdgeSampling - Edge from -14.098557027397089,136.0286709523649 to -13.853749011168235,136.18335264585954 is 31930m long, might cause issues with edge sampling.
``` 

Here are some of the ways where this applies:
- https://www.openstreetmap.org/way/29477902
- https://www.openstreetmap.org/way/731710835
- https://www.openstreetmap.org/way/544311447

My feeling would be, since there are not many of these edges it would be valid to simply ignore these cases and not interpolating there. I could no add a point for these ways and slightly improve the following of the original track, using satellite imagery, where I can see it, but it's probably not worth it. WDYT?

Reply:
Looking into those, the difference between the road that graphhopper shows with edge sampling the difference appears to be negligible.  I'll bump the threshold out to 1 degree instead of 1/4 degree, improve the error message,  and fall back to not interpolating in this case

Reply:
Ok, sounds good to me, thanks!

> Looking into those, the difference between the road that graphhopper shows with edge sampling the difference appears to be negligible.

How did you compare those? Visually using Leaflet? Leaflet by default uses a linear approach as well. There is a plugin for Leaflet: https://github.com/henrythasler/Leaflet.Geodesic

Or you could also use the GPX files and import them on Google MyMaps, they also use the great circle to connect two points.

Reply:
I switched to satellite view on local graphhopper web, the path was off by a couple of road widths in the middle.  I dug into the great circle approach, it wasn't too complex so I just implemented that for segments longer than 1/4 degree. Let me know what you think, it is based on the same source as `projectCoordinate` and seems to hold up pretty well in unit tests.